src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -45,19 +46,42 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
                         mpi_errno, "xpmem segmaps", MPL_MEM_SHM);
     for (i = 0; i < num_local; i++) {
         MPIDU_Init_shm_get(i, sizeof(xpmem_segid_t), &MPIDI_XPMEM_global.segmaps[i].remote_segid);
+        if (MPIDI_XPMEM_global.segmaps[i].remote_segid == -1) {
+            anyfail = true;
+        }
         MPIDI_XPMEM_global.segmaps[i].apid = -1;        /* get apid at the first communication  */
 
         /* Init AVL tree based segment cache */
         MPIDI_XPMEM_segtree_init(&MPIDI_XPMEM_global.segmaps[i].segcache);
     }
 
+    /* Check to make sure all processes initialized XPMEM correctly. */
+    if (anyfail) {
+        for (i = 0; i < num_local; i++) {
+            MPIDI_XPMEM_segtree_delete_all(&MPIDI_XPMEM_global.segmaps[i].segcache);
+        }
+        /* Not setting an mpi_errno value here because we can handle the failure of XPMEM
+         * gracefully. */
+        goto fn_fail;
+    }
```

Comment:
Just a minor comment: it might be more consistent if we handle failure for `segcache` and `segid` at the same place. E.g., release both at `fn_fail`.

Reply:
You mean `MPIDI_XPMEM_segtree_init`?

Reply:
No, I meant when `any_fail==true` you call `MPIDI_XPMEM_segtree_delete_all` here but call `xpmem_remove` at `fn_fail`. Both are for failure handling, so it is more consistent to put them at the same place.

Reply:
Oh. I did it that way because it's not as easy to tell if we should be calling `MPIDI_XPMEM_segtree_delete_all` inside `fn_fail`. Jumping there might have been triggered by something else before `MPIDI_XPMEM_segtree_init` is called.